runtime/commonMain/src/kotlinx/serialization/json/internal/StreamingJsonInput.kt

```diff
@@ -105,6 +106,12 @@ internal class StreamingJsonInput internal constructor(
         }
     }
 
+    private fun SerialDescriptor.getJsonElementIndex(key: String): Int {
+        if (!json.configuration.supportAlternativeNames) return this.getElementIndex(key)
```

Comment:
Please take into account the following invariants:
    * Fields rarely have alternative names even when this feature is enabled
    * Fields rarely have more than one alternative name

Then you can significantly speedup this method, just replace the nested map with two arrays (`String[]` and `int[]`), lookup with linear search and ensure that main name is always in the first position of the array

Reply:
Even if the property has one alternative name from `@JsonNames`, it means it has two names: one primary and one alternative, i.e., the map will contain two keys which point to the property's index. 

I don't quite get how it is possible to replace the map with two arrays. Put all names into one array, linearly search it, and then use index we found in the second array so it would give us the index of the field? I don't see how it is faster than a map, because even with disabled alternative names, `this.getElementName` still uses a map inside a `SerialClassDescImpl`. However, it is possible to do two-phase lookup: first for a primary name in the descriptor and second in the aforementioned array. But this case can still use a map – we just don't need to allocate it if a property does not have an alternative name.

Reply:
Good points, two-phase lookup seems to be slightly better.

Reply:
Just found out that this strategy would prevent us from detecting alternative name collisions with primary ones (because we do not build cache map until primary name is not found). So we either should choose speed or error detection